# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Sadaf-coder/flyrank-ml-work/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

Lane: Refresh / Content Opportunity Scoring (Lane 2)

I picked this lane because when I ran the starter notebook, I saw
something that surprised me — a simple hand-written rule only got
12 out of 50 pages right, but the random forest got 37 out of 50.
That gap felt worth exploring. I want to understand what signals
actually separate declining pages from stable ones, and build
something a real content team could use to decide where to spend
their time first.

In [2]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

My question: Which pages are most likely declining in search
visibility and should be looked at first?

Each row is one page. The output is a ranked list — highest risk
at the top. A content reviewer uses this list to decide which pages
to open and check. If I rank the wrong pages at the top, a reviewer
wastes their time. If I miss a truly declining page, it keeps losing
traffic while nobody notices — that second mistake feels worse to me.

I think ML can help here because the data showed real signal that
a simple rule couldn't fully capture.

In [3]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

**Supporting numbers from the starter dataset (30,000 pages):**

The following code cells load the starter data and show three real
numbers that support choosing this lane.

In [4]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

import os, sys, subprocess
import pandas as pd, numpy as np

# Setup
IN_COLAB = "google.colab" in sys.modules
REPO_DIR = "flyrank-ml-internship-starter"
if IN_COLAB and not os.path.isdir(REPO_DIR):
    subprocess.run(["git", "clone", "--depth", "1",
                    "https://github.com/flyrank-bih/flyrank-ml-internship-starter",
                    REPO_DIR], check=True)
if IN_COLAB:
    os.chdir(REPO_DIR)

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

# Number 1: How many pages are declining?
declining_rate = df["trend_direction"].str.lower().eq("down").mean()
print(f"1. Declining pages: {declining_rate:.1%} of all pages ({int(declining_rate*len(df)):,} of {len(df):,})")

# Number 2: Declining pages still have significant impressions
declining = df[df["trend_direction"].str.lower() == "down"]
median_imp = declining["impressions_90d"].median()
print(f"2. Median impressions (declining pages): {median_imp:,.0f} — these pages still matter")

# Number 3: Baseline vs ML model
print(f"3. Hand rule Precision@50: 0.240 (~12/50 correct)")
print(f"   Random forest Precision@50: 0.740 (~37/50 correct)")
print(f"   → ML is 3.1x better at finding the right pages to review")

1. Declining pages: 54.2% of all pages (16,262 of 30,000)
2. Median impressions (declining pages): 961 — these pages still matter
3. Hand rule Precision@50: 0.240 (~12/50 correct)
   Random forest Precision@50: 0.740 (~37/50 correct)
   → ML is 3.1x better at finding the right pages to review


From what I saw in the data, 54.2% of pages are declining and they
still get around 961 impressions — so these are not dead pages,
they still matter. The model found them better than my hand rule did.

But I can't say refreshing a page will fix it — I only observed a
pattern, I didn't run an experiment. And these results are on 30,000
rows, not the full dataset. I'll need to check again when I work
with the warehouse data.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

In [5]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.